pkg/cloudprovider/controller/nodecontroller.go

```diff
@@ -597,6 +597,13 @@ func (nc *NodeController) deletePods(nodeID string) error {
 		if pod.Status.Host != nodeID {
 			continue
 		}
+		if api.RestartPolicyNever == pod.Spec.RestartPolicy {
+			continue
+		} else if api.RestartPolicyOnFailure == pod.Spec.RestartPolicy {
+			if api.PodSucceeded == pod.Status.Phase {
```

Comment:
The right way to do this would be to add a job controller API to the system. #1624

There is a GSoC proposal for this. In the meantime, we're trying to make it simpler and more efficient to add a new controller to the system.